### DISCUSSION TITLE
Incorrect Identification of Twin Boundaries in MTEX for Copper Alpha Phase - MTEX 5.11.2

### DISCUSSION THREAD

**Discussion by victoriavolker** [mtex-toolbox/mtex#2283] [Ask Anything]
I am facing challenges in correctly identifying twin boundaries in my EBSD dataset using MTEX. The main issue is that the script identifies boundaries that are not twin boundaries as if they were.

I want to emphasize that, in my case:

- Twin boundaries only occur in the Copper Alpha phase.
- They are always parallel to each other, as defined by the Σ3 relationship and expected orientation.

Despite implementing criteria based on misorientation angles and a minimum boundary length, the code continues to include boundaries that should not be classified as twins. Adjusting thresholds or filtering by length has not resolved the issue. I believe the solution might require a more accurate approach to defining or verifying orientations.

Below is my full script for review and suggestions. I appreciate any insights or corrections that might help resolve this issue.

If further details about the dataset or parameters are needed, I am happy to provide them.

I'm using MTEX 5.11.2

```matlab
%% Implementation of Idea 2: Relative Analysis of Twin Boundary Lengths

% Define crystal symmetry for Copper and Zhanghenite phases
CS = {...
  'notIndexed', ...
  crystalSymmetry('m-3m', [3.6 3.6 3.6], 'mineral', 'Copper', 'color', [0.53 0.81 0.98]), ...
  crystalSymmetry('m-3m', [3 3 3], 'mineral', 'Zhanghenite', 'color', [0.56 0.74 0.56])};

% Set MTEX plotting conventions
setMTEXpref('xAxisDirection', 'east');  % Define x-axis pointing east
setMTEXpref('zAxisDirection', 'intoPlane');  % Define z-axis pointing into the plane

%% Specify File Names
pname = 'S:\02 HiWis\Victoria Köhler\12 Dissertation Brans\10 Materialcharakterisierung\Image Analysis';
fname = [pname '\AG451 6.crc'];  % File path for input data
![example](https://github.com/user-attachments/assets/71614517-d036-417d-8e4b-2b13b82bd504)
![example](https://github.com/user-attachments/assets/d2ad1618-a82e-479d-9530-acfc51c2bedc)

%% Import the Data
% Load EBSD data using the specified crystal symmetries and file interface
ebsd = EBSD.load(fname, CS, 'interface', 'crc', 'convertEuler2SpatialReferenceFrame');

% Segment grains based on orientation misorientation angle and minimum pixel size
[grains, ebsd.grainId, ebsd.mis2mean] = calcGrains(ebsd('indexed'), 'angle', 5*degree, 'minPixel', 10);

% Smooth grain boundaries for better visualization
grains = grains.smooth(5);

% Filter grains for the Copper phase only
grains = grains('Copper');

% Plot the filtered grains with their mean orientation
figure;
plot(grains, grains.meanOrientation);
title('Filtered Grains (Copper Phase)');

%% Define the Twinning Relationship Σ3 for Copper
CS = grains.CS;  % Get the crystal symmetry for Copper
gB = grains.boundary('Copper', 'Copper');  % Extract boundaries between Copper grains

% Define the twinning relationship based on a rotation axis and angle
h = Miller(-8, 7, 8, CS, 'uvw');  % Define the rotation axis
twinning = orientation('axis', h, 'angle', 59.6*degree, CS, CS);  % Create the twinning orientation

% Symmetrize the twinning relationship to capture all possible variants
twinning = symmetrise(twinning, CS);

% Identify boundaries that meet the twinning misorientation criteria
twinningAngleThreshold = 10*degree;  % Define threshold for twin boundary detection
isTwinning = angle(gB.misorientation, twinning) < twinningAngleThreshold;  % Identify twin boundaries

tB = gB(isTwinning);  % Extract twin boundaries

%% Analyze Twin Boundary Lengths
% Determine the maximum grain ID from the twin boundary data
maxGrainId = max(tB.grainId(:, 1));

% Compute total twin boundary length for each grain
componentLengths = accumarray(tB.grainId(:, 1), tB.segLength, [maxGrainId, 1], @sum, 0);

% Initialize an array to store the proportion of twin boundary lengths
grainsTwinLength = zeros(numel(grains), 1);

% Compute the proportion of twin boundary length relative to the grain perimeter
for i = 1:numel(grains)
    grainPerimeter = grains(i).perimeter;  % Total grain perimeter
    if grainPerimeter > 0 && i <= maxGrainId && ~isnan(componentLengths(i))
        grainsTwinLength(i) = componentLengths(i) / grainPerimeter;  % Twin boundary proportion
    end
end

%% Determine the Minimum Length Threshold
% Calculate a dynamic threshold based on mean and standard deviation of twin lengths
meanLength = mean(tB.segLength);
stdLength = std(tB.segLength);
minLengthThreshold = meanLength - 0.3 * stdLength;  % Less strict threshold
minLengthThreshold = max(0, minLengthThreshold);  % Ensure the threshold is positive

% Filter twin boundaries based on the minimum length threshold
isValidLength = tB.segLength > minLengthThreshold;
tB_filtered = tB(isValidLength);

%% Plot the Results
figure;
plot(grains, 'FaceAlpha', 0.5);  % Plot grains with transparency
hold on;

% Add grain boundaries
plot(grains.boundary, 'linecolor', 'k', 'linewidth', 0.5, 'displayName', 'Grain Boundary');

% Add filtered twin boundaries
if ~isempty(tB_filtered)
    twinLengths = tB_filtered.segLength;  % Twin boundary lengths
    plot(tB_filtered, twinLengths, 'linecolor', 'interp', 'linewidth', 2, ...
        'displayName', 'Filtered Twin Boundary');
    mtexColorbar('title', 'Filtered Twin Length');  % Color bar for twin lengths
else
    warning('No filtered twin boundaries identified.');
end

hold off;
legend show;  % Show legend
title('Filtered Twin Boundaries by Length');

```
I uploaded the image example. Marked in green is when identify correctly and in red is when identify incorrectly and on the ZIP file is the .crc  and .cpr file

![example](https://github.com/user-attachments/assets/2431f708-9ba5-4ae0-954b-9c607b2e93b7)
![phase](https://github.com/user-attachments/assets/7767da02-9cb8-4b40-8a06-5bbbc67b1596)
[AG451 6.zip](https://github.com/user-attachments/files/18210252/AG451.6.zip)

**Comment by kilir** [comment]
Hi,

in order to load the example data one also needs the cpr, not just the crc.
Also, could you explain why you do not like the boundaries encircled in red?

Cheers,

Rüdiger

**Reply by victoriavolker** [reply]
Hey,  

Thanks for your response! I added the `.cpr` to the script, but it doesn’t seem to have made any noticeable difference in the results.  

About the boundaries: the ones marked in red are incorrect because they’re just regular grain boundaries separating one grain from another, not twin boundaries. My goal is to accurately identify the twin boundaries so I can remove them from the image, leaving only the true grain boundaries. This is essential for my analysis since I need a clean dataset of actual grain divisions to perform a Neighbour Analysis later.  

By filtering out the twin boundaries, I want to focus only on the real grain structure without interference from features like twins. Any tips or suggestions on how I could tweak the script to better achieve this would be super helpful!  

Thanks in advance!

**Reply by kilir** [reply]
Hi,
no the .cpr doesn't change anything except that other people could try and load your data in order to help you with the problem - you only uploaded the .crc.

So maybe to rephrase: if it is not only misorientation (and I guess in your case you also want them to have a certain length), how would you tell a twin  from a grain boundary apart. What criterion do you have tio draw a red circle around the "bad" twin boundaries?

Cheers,

Rüdiger

**Comment by Tijmenvermeij** [comment]
Hi,

Two questions:

1. Why do you use the following code to define the twinning relation, and not just, for example, a 60deg rotation over the {111} axis? (To which your relation is very close...)
```
% Define the twinning relationship based on a rotation axis and angle
h = Miller(-8, 7, 8, CS, 'uvw');  % Define the rotation axis
twinning = orientation('axis', h, 'angle', 59.6*degree, CS, CS);  % Create the twinning orientation
```

2. Like Rüdiger, I wonder why you say that the red circled boundaries are wrongly identified?
